Closing a shell causes window to hang on OS X 10.11

There's several ways to get to this frozen shell problem since upgrading to 10.11. I don't know enough about C to help, but I would take a wild guess that it relates to OS X's new [SIPS.](https://en.wikipedia.org/wiki/System_Integrity_Protection).

- Quit a window with the key binding (e.g. in my case `C-d q`)
- Type exit at a shell
- Open the buffer mode pager thingy (e.g. in my case `C-d e` to open a scrollback buffer in less which I gather has something to do with copying and pasting but I just use it for scrolling)

This causes one window to basically freeze, In essence I can spawn windows but not close them.

On @martanne's advice I ran the following after intentionally causing the issue.

![screen shot 2016-02-18 at 12 31 57](https://cloud.githubusercontent.com/assets/2365737/13222281/2ae22e1e-d977-11e5-8e5a-fc2d63922dfe.png)

Which seemed to point to this line of code.

![screen shot 2016-02-18 at 12 31 29](https://cloud.githubusercontent.com/assets/2365737/13222285/2ed15db0-d977-11e5-842d-d7e809fbee22.png)

I've tried this with clean user configs (bashrc, profile etc), on two different machines and with clean installs of dvtm using the unmodified default config.

Hope someone can help because I've been using this thing daily for years now and I really miss it!

I'm probably not the best person to debug this.

From running `gcc dvtm-test -o dvtm-test`:

```
ld: warning: ignoring file dvtm-test, file was built for unsupported file format ( 0x23 0x64 0x65 0x66 0x69 0x6E 0x65 0x20 0x5F 0x50 0x4F 0x53 0x49 0x58 0x5F 0x43 ) which is not the architecture being linked (x86_64): dvtm-test
Undefined symbols for architecture x86_64:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

From `gcc dvtm-test.c -o dvtm-test`:

```
dvtm-test.c:44:15: error: use of undeclared identifier 'SIGWINCH'
    sigaction(SIGWINCH, &sa, NULL);
              ^
dvtm-test.c:48:26: error: use of undeclared identifier 'SIGWINCH'
    sigaddset(&blockset, SIGWINCH);
                         ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/usr/include/signal.h:122:52: note: expanded from macro 'sigaddset'
#define sigaddset(set, signo)   (*(set) |= __sigbits(signo), 0)
                                                     ^
2 errors generated.
```

--- a/dvtm.c
+++ b/dvtm.c
@@ -52,7 +52,7 @@ typedef struct {
 	int history;
 	int w;
 	int h;
-	volatile sig_atomic_t need_resize;
+	bool need_resize;
 } Screen;
 
 typedef struct {
@@ -248,6 +248,10 @@ static const char *shell;
 static Register copyreg;
 static volatile sig_atomic_t running = true;
 static bool runinall = false;
+static int sigwinch_pipe[] = {-1, -1};
+static int sigchld_pipe[] = {-1, -1};
+
+enum {PIPE_RD, PIPE_WR};
 
 static void
 eprint(const char *errstr, ...) {
@@ -698,6 +702,11 @@ get_client_by_coord(unsigned int x, unsigned int y) {
 
 static void
 sigchld_handler(int sig) {
+	write(sigchld_pipe[PIPE_WR], "\0", 1);
+}
+
+static void
+handle_sigchld() {
 	int errsv = errno;
 	int status;
 	pid_t pid;
@@ -731,6 +740,11 @@ sigchld_handler(int sig) {
 
 static void
 sigwinch_handler(int sig) {
+	write(sigwinch_pipe[PIPE_WR], "\0", 1);
+}
+
+static void
+handle_sigwinch() {
 	screen.need_resize = true;
 }
 
@@ -939,6 +953,14 @@ getshell(void) {
 	return "/bin/sh";
 }
 
+static bool
+set_blocking(int fd, bool blocking) {
+	int flags = fcntl(fd, F_GETFL, 0);
+	if (flags < 0) return false;
+	flags = blocking ? (flags & ~O_NONBLOCK) : (flags | O_NONBLOCK);
+	return !fcntl(fd, F_SETFL, flags);
+}
+
 static void
 setup(void) {
 	shell = getshell();
@@ -962,6 +984,23 @@ setup(void) {
 		colors[i].pair = vt_color_reserve(colors[i].fg, colors[i].bg);
 	}
 	resize_screen();
+
+	int *pipes[] = {&sigwinch_pipe[0], &sigchld_pipe[0]};
+	for (int i = 0; i < 2; ++i) {
+		int r = pipe(pipes[i]);
+		if (r < 0) {
+			perror("pipe()");
+			exit(EXIT_FAILURE);
+		}
+
+		for (int j = 0; j < 2; ++j) {
+			if (!set_blocking(pipes[i][j], false)) {
+				perror("fcntl()");
+				exit(EXIT_FAILURE);
+			}
+		}
+	}
+
 	struct sigaction sa;
 	memset(&sa, 0, sizeof sa);
 	sa.sa_flags = 0;
@@ -1815,19 +1854,12 @@ main(int argc, char *argv[]) {
 	KeyCombo keys;
 	unsigned int key_index = 0;
 	memset(keys, 0, sizeof(keys));
-	sigset_t emptyset, blockset;
 
 	setenv("DVTM", VERSION, 1);
 	if (!parse_args(argc, argv)) {
 		setup();
 		startup(NULL);
 	}
-
-	sigemptyset(&emptyset);
-	sigemptyset(&blockset);
-	sigaddset(&blockset, SIGWINCH);
-	sigaddset(&blockset, SIGCHLD);
-	sigprocmask(SIG_BLOCK, &blockset, NULL);
 
 	while (running) {
 		int r, nfds = 0;
@@ -1841,9 +1873,15 @@ main(int argc, char *argv[]) {
 		FD_ZERO(&rd);
 		FD_SET(STDIN_FILENO, &rd);
 
+		FD_SET(sigwinch_pipe[PIPE_RD], &rd);
+		nfds = MAX(nfds, sigwinch_pipe[PIPE_RD]);
+
+		FD_SET(sigchld_pipe[PIPE_RD], &rd);
+		nfds = MAX(nfds, sigchld_pipe[PIPE_RD]);
+
 		if (cmdfifo.fd != -1) {
 			FD_SET(cmdfifo.fd, &rd);
-			nfds = cmdfifo.fd;
+			nfds = MAX(nfds, cmdfifo.fd);
 		}
 
 		if (bar.fd != -1) {
@@ -1867,7 +1905,7 @@ main(int argc, char *argv[]) {
 		}
 
 		doupdate();
-		r = pselect(nfds + 1, &rd, NULL, NULL, NULL, &emptyset);
+		r = select(nfds + 1, &rd, NULL, NULL, NULL);
 
 		if (r < 0) {
 			if (errno == EINTR)
@@ -1901,6 +1939,18 @@ main(int argc, char *argv[]) {
 			}
 			if (r == 1) /* no data available on pty's */
 				continue;
+		}
+
+		if (FD_ISSET(sigwinch_pipe[PIPE_RD], &rd)) {
+			char buf[512];
+			while (read(sigwinch_pipe[PIPE_RD], &buf, sizeof(buf)) > 0);
+			handle_sigwinch();
+		}
+
+		if (FD_ISSET(sigchld_pipe[PIPE_RD], &rd)) {
+			char buf[512];
+			while (read(sigchld_pipe[PIPE_RD], &buf, sizeof(buf)) > 0);
+			handle_sigchld();
 		}
 
 		if (cmdfifo.fd != -1 && FD_ISSET(cmdfifo.fd, &rd))

